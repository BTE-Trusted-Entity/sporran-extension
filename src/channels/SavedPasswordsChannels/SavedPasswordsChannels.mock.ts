import { act } from '@testing-library/react';

import {
  hasSavedPasswordsChannel,
  getPasswordChannel,
} from './SavedPasswordsChannels';

jest.mock('./SavedPasswordsChannels');

const hasSavedPasswordsPromise = Promise.resolve(false);
jest
  .mocked(hasSavedPasswordsChannel.get)
  .mockReturnValue(hasSavedPasswordsPromise);

export async function waitForHasSavedPasswords(): Promise<void> {
  await act(async () => {
    await hasSavedPasswordsPromise;
  });
}

const getPasswordPromise = Promise.resolve('');
jest.mocked(getPasswordChannel.get).mockReturnValue(getPasswordPromise);

export async function waitForGetPassword(): Promise<void> {
  await act(async () => {
    await hasSavedPasswordsPromise;
  });
}
