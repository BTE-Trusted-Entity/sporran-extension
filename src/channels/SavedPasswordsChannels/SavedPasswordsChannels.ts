import { BrowserChannel } from '../base/BrowserChannel/BrowserChannel';
import { channelsEnum } from '../base/channelsEnum';

interface SavedPassword {
  password: string;
  timestamp: number;
}

const savedPasswords: Record<string, SavedPassword> = {};

interface SavePasswordInput {
  password: string;
  address: string;
}

export const savePasswordChannel = new BrowserChannel<SavePasswordInput>(
  channelsEnum.savePassword,
);

export async function savePassword({
  address,
  password,
}: SavePasswordInput): Promise<void> {
  savedPasswords[address] = { password, timestamp: Date.now() };
}

export const getPasswordChannel = new BrowserChannel<
  string,
  string | undefined
>(channelsEnum.getPassword);

export async function getPassword(
  address: string,
): Promise<string | undefined> {
  return savedPasswords[address]?.password;
}

export const forgetPasswordChannel = new BrowserChannel<string>(
  channelsEnum.forgetPassword,
);

export async function forgetPassword(address: string): Promise<void> {
  delete savedPasswords[address];
}

export const hasSavedPasswordsChannel = new BrowserChannel<void, boolean>(
  channelsEnum.hasSavedPasswords,
);

export async function hasSavedPasswords(): Promise<boolean> {
  return Object.values(savedPasswords).length > 0;
}

export const forgetAllPasswordsChannel = new BrowserChannel(
  channelsEnum.forgetAllPasswords,
);

export async function forgetAllPasswords(): Promise<void> {
  for (const password in savedPasswords) {
    delete savedPasswords[password];
  }
}

const saveDuration = 15 * 60 * 1000;
const intervalDuration = 1 * 60 * 1000;

function checkExpiredPasswords(): void {
  const oldestPossible = Date.now() - saveDuration;
  for (const password in savedPasswords) {
    const { timestamp } = savedPasswords[password];
    if (timestamp < oldestPossible) {
      delete savedPasswords[password];
    }
  }
}

export function schedulePasswordsCheck(): void {
  setInterval(checkExpiredPasswords, intervalDuration);
}
