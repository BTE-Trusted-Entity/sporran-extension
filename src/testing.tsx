export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
