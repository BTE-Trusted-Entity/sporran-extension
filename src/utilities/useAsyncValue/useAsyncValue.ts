import { useEffect, useState } from 'react';

import { exceptionToError } from '../exceptionToError/exceptionToError';

export function useAsyncValue<Input extends Array<unknown>, Output>(
  fetcher: (...args: Input) => Promise<Output>,
  args: Input,
): Output | undefined {
  const [data, setData] = useState<Output>();
  const [error, setError] = useState<Error>();

  useEffect(() => {
    (async () => {
      try {
        setData(await fetcher(...args));
      } catch (exception) {
        console.error(exception);
        setError(exceptionToError(exception));
      }
    })();
  }, args); // eslint-disable-line react-hooks/exhaustive-deps

  if (error) {
    throw error;
  }

  return data;
}
