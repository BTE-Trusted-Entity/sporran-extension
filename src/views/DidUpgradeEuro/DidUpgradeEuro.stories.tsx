import { Meta } from '@storybook/react';

import { identitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';

import { DidUpgradeEuro } from './DidUpgradeEuro';

export default {
  title: 'Views/DidUpgradeEuro',
  component: DidUpgradeEuro,
} as Meta;

export function Template(): JSX.Element {
  return (
    <DidUpgradeEuro
      identity={identities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']}
    />
  );
}
