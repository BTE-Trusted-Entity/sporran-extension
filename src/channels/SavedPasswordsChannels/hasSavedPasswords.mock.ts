import { act } from '@testing-library/react';

import { hasSavedPasswordsChannel } from './SavedPasswordsChannels';

jest.mock('./SavedPasswordsChannels');
const hasSavedPasswordsPromise = Promise.resolve(false);
(hasSavedPasswordsChannel.get as jest.Mock).mockReturnValue(
  hasSavedPasswordsPromise,
);

export async function waitForHasSavedPasswords(): Promise<void> {
  await act(async () => {
    await hasSavedPasswordsPromise;
  });
}
