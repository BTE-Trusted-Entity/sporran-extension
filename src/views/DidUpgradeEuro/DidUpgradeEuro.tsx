import { browser } from 'webextension-polyfill-ts';

import { Link } from 'react-router-dom';

import { FormEvent, useCallback } from 'react';

import * as styles from './DidUpgradeEuro.module.css';

import { Identity } from '../../utilities/identities/types';
import { IdentitySlide } from '../../components/IdentitySlide/IdentitySlide';
import { paths } from '../paths';
import {
  PasswordField,
  usePasswordField,
} from '../../components/PasswordField/PasswordField';
import { LinkBack } from '../../components/LinkBack/LinkBack';
import { getTransaction } from '../../utilities/didUpgrade/didUpgrade';

import { useAsyncValue } from '../../utilities/useAsyncValue/useAsyncValue';
import { useTXDSubmitter } from '../../utilities/useTXDSubmitter/useTXDSubmitter';
import {
  getIdentityCryptoFromSeed,
  getLightDidFromSeed,
} from '../../utilities/identities/identities';
import {
  getCheckoutURL,
  getCheckoutCosts,
} from '../../utilities/checkout/checkout';

interface Props {
  identity: Identity;
}

export function DidUpgradeEuro({ identity }: Props): JSX.Element | null {
  const t = browser.i18n.getMessage;

  const { address } = identity;

  const passwordField = usePasswordField();

  const cost = useAsyncValue(getCheckoutCosts, [])?.did;
  const submitter = useTXDSubmitter();

  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();

      if (!submitter) {
        throw new Error('Submitter address missing');
      }

      const { seed } = await passwordField.get(event);
      const { keypair, sign } = await getIdentityCryptoFromSeed(seed);

      const document = getLightDidFromSeed(seed);
      const { extrinsic } = await getTransaction(
        document,
        keypair,
        sign,
        submitter,
      );

      const checkout = await getCheckoutURL();

      const url = new URL(checkout);
      url.searchParams.set('address', address);
      url.searchParams.set('tx', extrinsic.method.toHex());

      await browser.tabs.create({ url: url.toString() });
      window.close();
    },
    [address, passwordField, submitter],
  );

  if (!cost) {
    return null; // network data pending
  }

  return (
    <form
      onSubmit={handleSubmit}
      className={styles.container}
      autoComplete="off"
    >
      <h1 className={styles.heading}>{t('view_DidUpgradeEuro_heading')}</h1>
      <p className={styles.subline}>{t('view_DidUpgradeEuro_subline')}</p>

      <IdentitySlide identity={identity} />

      <dl className={styles.details}>
        <div className={styles.detail}>
          <dt className={styles.detailName}>
            {t('view_DidUpgradeEuro_address')}
          </dt>
          <dd className={styles.detailValue}>{address}</dd>
        </div>
      </dl>

      <PasswordField identity={identity} autoFocus password={passwordField} />

      <p className={styles.info}>
        <span>{t('view_DidUpgradeEuro_info')}</span>
        <span className={styles.paypal}>{t('view_DidUpgradeEuro_paypal')}</span>
        <span>{t('view_DidUpgradeEuro_cost', [cost])}</span>
      </p>

      <p className={styles.buttonsLine}>
        <Link to={paths.home} className={styles.cancel}>
          {t('common_action_cancel')}
        </Link>
        <button type="submit" className={styles.submit}>
          {t('common_action_sign')}
        </button>
      </p>

      <LinkBack />
    </form>
  );
}
