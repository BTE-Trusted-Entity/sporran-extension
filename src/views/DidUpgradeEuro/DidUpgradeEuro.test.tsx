import { identitiesMock as identities, render } from '../../testing/testing';

import { waitForGetPassword } from '../../channels/SavedPasswordsChannels/SavedPasswordsChannels.mock';

import { useAsyncValue } from '../../utilities/useAsyncValue/useAsyncValue';

import { DidUpgradeEuro } from './DidUpgradeEuro';

jest.mock('../../utilities/useAsyncValue/useAsyncValue');
jest.mocked(useAsyncValue).mockReturnValue({ did: 'EUR 4,00' });

describe('DidUpgradeEuro', () => {
  it('should render', async () => {
    const { container } = render(
      <DidUpgradeEuro
        identity={
          identities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']
        }
      />,
    );
    await waitForGetPassword();
    expect(container).toMatchSnapshot();
  });
});
