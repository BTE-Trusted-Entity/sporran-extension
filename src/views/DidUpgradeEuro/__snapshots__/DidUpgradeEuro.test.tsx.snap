// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DidUpgradeEuro should render 1`] = `
<div>
  <form
    autocomplete="off"
    class="container"
  >
    <h1
      class="heading"
    >
      Prepare Transaction
    </h1>
    <p
      class="subline"
    >
      Get your DID
    </p>
    <section
      class="container"
    >
      <div
        class="avatar"
      >
        Identicon
      </div>
      <div
        class="centeredNameLine"
      >
        <span
          class="name"
        >
          Light DID Identity
        </span>
      </div>
    </section>
    <dl
      class="details"
    >
      <div
        class="detail"
      >
        <dt
          class="detailName"
        >
          Account Address:
        </dt>
        <dd
          class="detailValue"
        >
          4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1
        </dd>
      </div>
    </dl>
    <p
      class="resetLine"
    >
      <label
        class="passwordLabel"
        for="password"
      >
        Sign with password:
      </label>
      <a
        class="reset"
        href="/identity/4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1/reset"
      >
        Reset password
      </a>
    </p>
    <p
      class="passwordLine"
    >
      <input
        class="password"
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button
        aria-label="Show"
        class="show"
        title="Show"
        type="button"
      />
      <output
        class="errorTooltip"
        hidden=""
      />
    </p>
    <label
      class="rememberLabel"
    >
      <span>
        Remember password for the next 15 minutes
      </span>
      <input
        class="remember"
        name="remember"
        type="checkbox"
      />
      <span />
    </label>
    <p
      class="info"
    >
      <span>
        After signing you will go to the checkout page to pay with 
      </span>
      <span
        class="paypal"
      >
        PayPal 
      </span>
      <span>
        (total cost: EUR 4,00).
      </span>
    </p>
    <p
      class="buttonsLine"
    >
      <a
        class="cancel"
        href="/"
      >
        Cancel
      </a>
      <button
        class="submit"
        type="submit"
      >
        Sign
      </button>
    </p>
    <button
      aria-label="Back"
      class="linkBack"
      title="Back"
      type="button"
    />
  </form>
</div>
`;
